terps/unp64/unp64.cpp

```diff
@@ -0,0 +1,745 @@
+/* This is a cut-down version of UNP64 with only the bare minimum
+ * needed to decompress a number of Commodore 64 adventure games.
+ * It is distributed under the zlib License by kind permission of
+ * the original authors Magnus Lind and iAN CooG.
+ */
+
+/*
+UNP64 - generic Commodore 64 prg unpacker
+(C) 2008-2022 iAN CooG/HVSC Crew^C64Intros
+original source and idea: testrun.c, taken from exo20b7
+
+Follows original disclaimer
+*/
+
+/*
+ * Copyright (c) 2002 - 2023 Magnus Lind.
+ *
+ *
+ * This software is provided 'as-is', without any express or implied
+ * warranty. In no event will the authors be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ * claim that you wrote the original software. If you use this software
+ * in a product, an acknowledgment in the product documentation would be
+ * appreciated but is not required.
+ * 2. Altered source versions must be plainly marked as such, and must not be
+ * misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source distribution.
+ *
+ */
+
+/*
+ C++ version based on code adapted to ScummVM by Avijeet Maurya
+ */
+
+
+#include <cstring>
+#include <cstdio>
+
+#include "types.h"
+#include "globals.h"
+#include "6502_emu.h"
+#include "exo_util.h"
+#include "unp64.h"
+
+#include "unp64_interface.h"
+
+namespace Unp64 {
+
+void reinitUnp(void) {
+	_G(_unp)._idFlag = 0;
+	_G(_unp)._forced = 0;
+	_G(_unp)._strMem = 0x800;
+	_G(_unp)._retAdr = 0x800;
+	_G(_unp)._depAdr = 0;
+	_G(_unp)._endAdr = 0x10000;
+	_G(_unp)._rtAFrc = 0;
+	_G(_unp)._wrMemF = 0;
+	_G(_unp)._lfMemF = 0;
+	_G(_unp)._exoFnd = 0;
+	_G(_unp)._ecaFlg = 0;
+	_G(_unp)._fEndBf = 0;
+	_G(_unp)._fEndAf = 0;
+	_G(_unp)._fStrAf = 0;
+	_G(_unp)._fStrBf = 0;
+	_G(_unp)._mon1st = 0;
+}
+
+int isBasicRun1(int pc) {
+	if (pc == 0xa7ae || pc == 0xa7ea || pc == 0xa7b1 || pc == 0xa474 || pc == 0xa533 || pc == 0xa871 || pc == 0xa888 || pc == 0xa8bc)
+		return 1;
+	else
+		return 0;
+}
+
+int isBasicRun2(int pc) {
+	if (isBasicRun1(pc) || ((pc >= 0xA57C) && (pc <= 0xA659)) || pc == 0xa660 || pc == 0xa68e)
+		return 1;
+	else
+		return 0;
+}
+
+int unp64cpp(uint8_t *compressed, size_t length, uint8_t *destinationBuffer, size_t *finalLength, const char *switches) {
+
+    char settings[4][64];
+    int numSettings = 0;
+
+    if (switches != NULL) {
+        Unp64::size_t string_length = std::strlen(switches);
+        char string[100];
+        if (string_length > 0 && string_length <= 100) {
```

Comment:
This will pass if `string_length` is 100, and then effectively do this on line 98: `string[100] = '\0'`; but `string` is only 100 bytes long, so that will overflow it.